lib/component.js

```diff
@@ -616,12 +623,21 @@ class Component extends WebComponent {
       const attrType = attrSchema.type;
       let attrValue = null;
 
+      if (attrSchema.deprecatedMsg) {
+        console.warn(`${this}: attr '${attr}' is deprecated. ${attrSchema.deprecatedMsg}`);
+      }
+
       if (!this.hasAttribute(attr)) {
-        if (attrType === `boolean` && attrSchema.default) {
-          console.warn(
-            `${this}: attr '${attr}' is defaulted to true. Default-true boolean attributes are deprecated and will be removed in a future release`,
+        if (attrType === `boolean` && (attrSchema.default || attrsSchema.required)) {
+          throw new Error(
+            `${this}: boolean attr '${attr}' cannot have required or default, since its value is derived from whether dom element has the attribute, not its value`,
           );
         }
+
+        if (attrSchema.required) {
```

Comment:
could write this as
```
if(!attrSchema.required) {
  attrValue = attrSchema.default;
}
```

Reply:
We'd ultimately then assign null to `this._attrs[attr]`

Reply:
talked offline, I think it's ambiguous to have both required and `!this.hasAttribute(attr)`. Rather than return silently, this should throw an Error

Reply:
This is run during the constructor, in which case the attribute may not have been set yet, so we can't error at this point 😞

Reply:
yeah we gotta do the check in connectedCallback or whatever that calls